Added toolbar dropdown action to set any image as feature image

up1512001 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
- if the feature image is not set for the post then if someone checked the toggle control of the image to set it as a feature image then it will be set as a feature image for the post.

## Why?
fixes #65889 

## How?
- added toggle control in inspector control to `Set this image as feature image`.
- if `isFeatureImage` attribute is set & post doesn't have a feature image then it will be set as a feature image for the post.

## Testing Instructions
- create a new post.
- add image block, and check `Set this image as feature image` toogle control of settings to set this image as feature image.
- check feature image of post.

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/4576c7c0-61df-43fd-a062-c47a6029bbe5

up1512001 committed:
feature: created toggle control to set content image as feature image if no feature image is set

up1512001 committed:
Fix: docs build

paaljoachim commented:
I will buzz the Gutenberg designers here: @WordPress/gutenberg-design 
To get feedback on this PR.

t-hamano commented:
Thanks for the PR!

There are a few concerns with the approach to running `set_post_thumbnail()` dynamically:

- This will not work if the image source is a URL instead of media.
- This will not work if the image is placed inside a query loop or in a single template.
- If there are multiple images with "Set this image as feature image" enabled in the content, the first image will be set as the featured image, but the settings for the second and subsequent images will be ignored.
- The featured image will not update until the image is actually rendered, i.e. until you save the post.

I think of this feature as an action button, not a toggle, so it should work like this:

- The "Make featured image" button will only be displayed in the block toolbar if the block has access to the post context, i.e. inside the post editor or the query loop block.
- When the button is clicked:
  - If the image source is media, it will _actually_ update the featured image based on that media.
  - If the image source is a URL, it will download the image and upload it to media. Then it will actually update the featured image based on that media. TheIimage block's source will also be updated from URL to media.
- The user will be notified via the snackbar when the featured image has been updated.

I need some design feedback, but I'm thinking of something like this:

![image](https://github.com/user-attachments/assets/0f99b497-0a48-4519-8bad-da5208381ae1)

jameskoster commented:
Providing a way to set the featured image via compatible image blocks seems like a good idea to explore. I agree with @t-hamano that a toggle in the Inspector doesn't quite work, and the reasoning for placing this in the block toolbar. 

However the prominence in the proposed design seems a bit too much, given it would be visible on every single image block.

I'm also curious what happens after you make the image featured. Is the link between the original block and the featured image retained? IE if you subsequently change the featured image does the original image block update, or is it a one-way interaction?

t-hamano commented:
My guess is it's one-way - we just push it from the Image block to the featured image. We could probably link it too, but that might get a bit more complicated.

> However the prominence in the proposed design seems a bit too much, given it would be visible on every single image block.

Yes, I've tried to find a suitable icon to replace it, but I haven't found one yet 😅

t-hamano commented:
> What happens if one clicks another Image block star icon? Will the first Image block star icon be deactivated?

If the approach is to label the Image block as a featured image, then when one Image block's icon is pressed, I think the icons of other Image blocks should be disabled.

On the other hand, if we take the approach of pushing the current image to a featured image, the icon should remain available in all image blocks since it represents an action.

> I believe there is an issue for adding first image in post content as featured in the Global style setting.

Perhaps #65107? If we could move forward with #65107, would it satisfy the needs of our customers?

t-hamano commented:
If we were to use an icon, I would imagine something like this (sorry for the crappy icon!):

![image](https://github.com/user-attachments/assets/db7c0d1c-b948-4eb3-8607-232137b2aea5)

jameskoster commented:
I'm wondering if even the icon is too prominent. While handy, this doesn't feel like an affordance users will use all that often? Would it make sense in the ellipsis menu?

t-hamano commented:
> Would it make sense in the ellipsis menu?

This is possible. It looks like there are two slots in the ellipsis menu, and we can inject a button at the top, or between "Create pattern" and "Edit as HTML":

| Header | Header |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/bad5af40-a284-4710-bb73-fef4bfc123c1)| ![image](https://github.com/user-attachments/assets/f370637e-3b47-47bb-bd8c-02ec533459dd)|

up1512001 commented:
@t-hamano @jameskoster I think if we add this kind of toolbar control it will be much better. let me know your view.

https://github.com/user-attachments/assets/8a4e122d-e295-4620-8267-0459af05bdcd

up1512001 committed:
revert: php changes for feature image

up1512001 committed:
feature: created toolbar control to set image as feature image is post don't have feature image set

jameskoster commented:
@up1512001 Generally speaking the toolbar is a location reserved for objectively the most common actions users might apply. Ideally they are carefully curated to avoid an overwhelming/confusing number of options.

Over the course of a post or pages lifetime how often do we expect users to make use of this feature? In most cases it will be once, and that's assuming they don't set the featured image via the dedicated control. For that reason, if we add this feature I'd prefer to see it in the ellipsis menu. Another benefit is that the action becomes available in List view.

@t-hamano of those options I think the second one makes most sense.

t-hamano commented:
@up1512001 We should be able to inject buttons into the ellipsis menu via the `BlockSettingsMenuControls` slot. See [the implementation in the Template Part block](https://github.com/WordPress/gutenberg/blob/d6a4c72606d97d4b9e1593cce28f6bf9d5448052/packages/block-library/src/template-part/edit/index.js#L282-L308) for reference.

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -1043,10 +1047,45 @@ export default function Image( {
+	const makeFeatureImageFormatControl = (
+		<BlockControls>
+			<ToolbarGroup>
+				<ToolbarButton
+					icon="format-image"
+					label={ __( 'Make Feature Image' ) }
+					onClick={ () =>
+						setAttributeForFeatureImage( isFeatureImage )
+					}
+					isPressed={ isFeatureImage }
+				/>
```
@paaljoachim @jameskoster  @scruffian 

I want to confirm again here: does the "Make Feature Image" button mean to _label_ the block to be used as a featured image? Or is it meant to _update_ the featured image with the current image?

For example, let's assume that the former (labeling) is what is expected here. Furthermore, suppose that [useFirstImageFromPost](https://github.com/WordPress/gutenberg/pull/56573) is enabled for the featured image and a second Image block in the content is labeled to be used as a featured image.

In this case, which one will actually be displayed as the featured image?

To prevent such inconsistencies, I prefer the latter approach (updating the featured image with the current image).

jameskoster replied:
The latter was my expectation. But I don't have a strong feeling about this feature and would welcome more feedback.

paaljoachim replied:
I expect to use the image as a selected featured image.

That is when I press the link/button "Set featured image" associated with the Image block a featured image should also be added here: 
![Screenshot 2024-10-08 at 20 34 13](https://github.com/user-attachments/assets/668b079a-75fb-409d-83bf-8a447b1b16e5)

It becomes an additional option to directly set the Featured image through the ellipse (3 button drop down) in an Image block. As it is added here it could easily also be added to a Cover block, Media & Text block etc.
If there was already a featured image added to the post then clicking the option in the ellipse would then updated the featured image with the one that was just selected through the Image block.

Conclusion is that having an additional method to add a featured image would be helpful. 

--

Now on another similar issue.
It would also be helpful to set first image in post as featured image through Global settings. This means if one forgets to set a featured image then the first image in the post will automatically be set as featured. I have experienced over and over that customers forget to add a featured image. Having this happen automatically would very much help.

up1512001 committed:
feature: created block settings & added proper notices for success message

up1512001 commented:
@t-hamano @jameskoster I have added `Make Feature Image` option on image toolbar ellipsis menu, please check this 🙇 

https://github.com/user-attachments/assets/aec85b98-c45d-4905-9fa5-fac0392c3a9c

up1512001 committed:
Merge branch 'trunk' into gh/65889

paaljoachim commented:
It looks good @up1512001 

You can use the same language "Set featured image" as is used in the post settings area for the featured image.

t-hamano commented:
@up1512001 Thanks for the update!

We need to improve the code and address more scenarios to move this PR forward.

Here's what I can think of for now:

- The block attribute, i.e. `isFeaturedImage`, should no longer be necessary.
- This function should not be available if we can't reference the post id or if we're in a query loop. For example, [the Featured Image has a check for those](https://github.com/WordPress/gutenberg/blob/5529e44dbf8b2ae13e9837c2f122458caf6a2c40/packages/block-library/src/post-featured-image/edit.js#L249-L258). This function should have a similar check.
- We should use `useEntityProp` instead of `dispatch`. Additionally, using the `core/editor` store is not recommended (see [this example](https://github.com/WordPress/gutenberg/blob/5529e44dbf8b2ae13e9837c2f122458caf6a2c40/packages/block-library/src/template-part/edit/advanced-controls.js#L58-L62)). Here's [a good reference](https://github.com/WordPress/gutenberg/blob/5529e44dbf8b2ae13e9837c2f122458caf6a2c40/packages/block-library/src/post-featured-image/edit.js#L75-L80) on how to update the featured image via `useEntityProp`.
- The image source may be a URL instead of a media. If the image source is a URL, this button will not work. If we also want to consider the URL source, we need to take care of uploading the image to the media beforehand. Personally, I think this PR should only consider the media source, and address the URL source in a follow-up.

If you have any questions, please feel free to contact us 👍

up1512001 committed:
Merge branch 'trunk' into gh/65889

up1512001 committed:
update: added required block context for feature image control

up1512001 committed:
update: set feature image control as per suggestion

up1512001 committed:
remove: unnessary isFeature image attribute

up1512001 committed:
rename: feature image control

up1512001 committed:
Fix: query loop issue

up1512001 committed:
Fix: typo in set featured image

up1512001 commented:
@t-hamano Addressed your feedback please review this PR.

- if an image block is added into the query loop then the `Set featured image` control will not be available.
- Also if the current image is set as a feature image then this control won't be visible.


https://github.com/user-attachments/assets/87d4c8a4-babf-4451-a901-6a9e4d64dae9


cc: @paaljoachim

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -1043,10 +1061,43 @@ export default function Image( {
+				( selectedClientIds.length === 1 &&
+					! isDescendentOfQueryLoop &&
+					postId &&
+					storedFeaturedImage !== id &&
```
> Also if the current image is set as a feature image then this control won't be visible.

This conditional statement creates an accessibility issue.

When keyboard users click the button, they lose focus because the button disappears. For now, we can simply remove the condition (`storedFeaturedImage !== id`).

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -1043,10 +1061,43 @@ export default function Image( {
+	 * Set the post's featured image with the current image.
+	 */
+	const setPostFeatureImage = () => {
+		setFeaturedImage( id );
```
```suggestion
		setFeaturedImage( id );
		createSuccessNotice( __( 'Post featured image updated.' ), {
			type: 'snackbar',
		} );
```

It would be good to give a notice here, the text may have room for improvement.

![f64055e3940f0ae33c64b4dff0b7fb57](https://github.com/user-attachments/assets/d3a74b8a-3ea0-41ee-9e4f-9690cd15609f)

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -1043,10 +1061,43 @@ export default function Image( {
+					postId &&
+					storedFeaturedImage !== id &&
+					clientId === selectedClientIds[ 0 ] && (
+						<MenuItem onClick={ () => setPostFeatureImage() }>
```
```suggestion
						<MenuItem onClick={ setPostFeatureImage }>
```

Nit

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -1043,10 +1061,43 @@ export default function Image( {
 		<>
 			{ mediaReplaceFlow }
 			{ controls }
+			{ featureImageControl }
```
```suggestion
			{ featuredImageControl }
```

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -1043,10 +1061,43 @@ export default function Image( {
+		setFeaturedImage( id );
+	};
+
+	/**
+	 * Create a menu item to set the image as the post's featured image.
+	 *
+	 * @param {Object} props                   - Props for block settings menu control.
+	 * @param {Object} props.selectedClientIds - Selected client ids.
+	 *
+	 * @return {void} - Menu item to set the image as the post's featured image.
+	 */
```
This component doesn't have any props, and since it's an internal component, I personally don't think it needs JSDoc at all.

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -1043,10 +1061,43 @@ export default function Image( {
+						<MenuItem onClick={ () => setPostFeatureImage() }>
+							{ __( 'Set featured image' ) }
+						</MenuItem>
+					) ) ||
+				null
```
I don't think `|| null` is necessary.

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -842,6 +844,22 @@ export default function Image( {
+	const { postType: postTypeSlug, postId, queryId } = context;
+	const isDescendentOfQueryLoop = Number.isFinite( queryId );
+
+	/**
+	 * Get feature image from current post.
+	 */
```
```suggestion
```

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -842,6 +844,22 @@ export default function Image( {
 	const shadowProps = getShadowClassesAndStyles( attributes );
 	const isRounded = attributes.className?.includes( 'is-style-rounded' );
 
+	/**
+	 * Get the post type, post ID, and query ID from the context for featured image control.
+	 */
+	const { postType: postTypeSlug, postId, queryId } = context;
```
```suggestion
	const { postType, postId, queryId } = context;
```

- This comment seems redundant to me, since it's pretty obvious what this code does.
- Since there are no conflicting variable names, I don't think we need to use aliases.

t-hamano reviewed (commented):
Thanks for the update! This PR is working fine. It even works in the Pages screen of the site editor:

https://github.com/user-attachments/assets/57f615ed-4ca8-4093-9912-db3c3aeca5bf


As suggested in [the comment](https://github.com/WordPress/gutenberg/pull/65896#issuecomment-2401187964), this doesn't work if the image source is a URL. I think we need to check `attributes.id` for now.

up1512001 committed:
Merge branch 'trunk' into gh/65889

up1512001 committed:
update as per suggestions

up1512001 commented:
@t-hamano updated as per your suggestions 🙇‍♂️ 
- Added `attributes.id` check if image source is URL.
- added snackbar success message.
- removed check that was hiding the `Set featured image` option if the current image is the featured image.

https://github.com/user-attachments/assets/8cd695bc-6566-491e-b4df-5e50e669e5f8

t-hamano commented on packages/block-library/src/image/image.js:
```diff
@@ -1094,10 +1105,38 @@ export default function Image( {
+				selectedClientIds.length === 1 &&
+				! isDescendentOfQueryLoop &&
+				postId &&
+				undefined !== id &&
+				null !== id &&
```
```suggestion
				id &&
```

Nit:

up1512001 replied:
@t-hamano Fixed.

t-hamano reviewed (commented):
Thanks for the update! From a code perspective, it looks almost ready to ship.

@jameskoster @paaljoachim Any feedback on the copy of the notice below?

![f64055e3940f0ae33c64b4dff0b7fb57](https://github.com/user-attachments/assets/d3a74b8a-3ea0-41ee-9e4f-9690cd15609f)

up1512001 committed:
Fix: minor feedback

up1512001 committed:
Fix: minor issue

paaljoachim commented:
I got to say that this is a really nice job @up1512001 Utsav! Very nice!

The notification message looks good. I do not know if it is a similar syntex (kind) as used with other notifications so we will see what Jay says.

t-hamano reviewed (approved):
Ok, let's ship this PR. As for the notification message, if there is a more reasonable one we can address that in a follow-up.

t-hamano commented:
@up1512001 Could you merge the trunk branch into this PR again? That should solve the stalled CI (See #66771).

up1512001 committed:
Merge branch 'trunk' into gh/65889
